keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -0,0 +1,656 @@
+   - `protocolSignature`: Specifies the protocol targeted by this Bucket instance.  One of:
+     - `azureBlob`: data required to target a provisioned azure container and/or storage account.
+     - `s3`: data required to target a provisioned S3 bucket and/or user.
+     - `gcs`: data required to target a provisioned GCS bucket and/or service account.
+10. `allowedNamespaces`: a copy of the `BucketClass`'s allowed namespaces. Additionally, this list can be mutated by the admin to allow or deny namespaces over the life of the bucket.
+11. `parameters`: a copy of the BucketClass parameters.
+12. `conditions`: the latest available observation of this `Bucket`. (Phase is not used.)
+13. `Type`: only one type is supported: "BucketAvailable". If the condition's `Status` is true then the bucket is available to be accessed for the `Reason` provided. If `Status` is false then the bucket is unable to be accessed and, again, `Reason` states why.
+14. `Status`: one of true, false or unknown. If not specified then unknown is assumed. A true status indicates the this condition is true at the time of observation.
+15. `Reason`: one of "Available", "Bound", "Creating", "Deleting", "Released". Note: "Available" indicates the `Bucket` was observed and its `conditions` was empty, which is expected when a brownfield `Bucket` is manually created.
+16. `Message`: optional human readable message containing more details.
+
+#### BucketClass
+
+A cluster-scoped custom resource to provide admins control over the handling of bucket provisioning.  The `BucketClass` (BC) defines a release policy, specifies driver specific parameters, and provides the provisioner name. A list of allowed namespaces can be specified to restrict new bucket creation and access to existing buckets. A default bucket class can be defined for each supported protocol, which allows the bucket class to be omitted from a `BucketRequest`. Most of the `BucketClass` fields are copied to the generated `Bucket` instance.
+
+> Note: a `BucketClass` is immutable, like a storage class, but the fields copied to the `Bucket` can be edited by the admin.
+
+```yaml
+apiVersion: cosi.io/v1alpha1
+kind: BucketClass
+metadata:
+  name: 
+provisioner: [1]
+isDefaultBucketClass: [2]
+protocol: {"azureblob", "gcs", "s3", ... } [3]
+anonymousAccessMode: [4]
+releasePolicy: {"Delete", "Retain"} [5]
+allowedNamespaces: [6]
+  - name:
+    uid:
+parameters: [7]
+```
+
+1. `provisioner`: the name of the driver. If supplied the driver container and sidecar container are expected to be deployed. Format: <provisioner-namespace>"/"<provisioner-name>, eg "ceph-rgw-provisoning/ceph-rgw.cosi.ceph.com".
+2. `isDefaultBucketClass`: (optional) boolean, default is false. If set to true then potentially a `BucketRequest` does not need to specify a `BucketClass`. If the greenfield `BucketRequest` omits the `BucketClass` and a default `BucketClass`'s protocol matches the `BucketRequest`'s protocol then the default bucket class is used.
```

Comment:
What happens when there are two default `BucketClass`es for the same `protocol`? If this is never supposed to happen, who (which admission controller, I guess) is responsible for this to be the case?

Reply:
The COSI central controller will register webhooks for this. This is not currently shown since we want to minimize the number of changes before the next KEP review. This adminssion web hook does not affect the overall architecture.

Reply:
My point is mainly: the spec ought to define how the system behaves in various states of the system, or which invariants are to be kept (e.g., only a single default BC per protocol). Indeed, how this is enforced in the implementation is a design decision and should not necessarily be defined in this spec.